Code_Exercises/Exercise_11_In_Order_Queue/solution.cpp

```diff
@@ -0,0 +1,211 @@
+
+      cgh.parallel_for<kernel_c_1>(sycl::range{dataSize}, [=](sycl::id<1> idx) {
+        accOut[idx] -= accIn[idx];
+      });
+    });
+
+    inOrderQueue.submit([&](sycl::handler& cgh) {
+      auto accInA = bufInB.get_access<sycl::access::mode::read>(cgh);
+      auto accInB = bufInC.get_access<sycl::access::mode::read>(cgh);
+      auto accOut = bufOut.get_access<sycl::access::mode::write>(cgh);
+
+      cgh.parallel_for<kernel_d_1>(sycl::range{dataSize}, [=](sycl::id<1> idx) {
+        accOut[idx] = accInA[idx] + accInB[idx];
+      });
+    });
+
+    inOrderQueue.wait_and_throw();
+  } catch (const sycl::exception& e) {
+    std::cout << "Exception caught: " << e.what() << std::endl;
+  }
+
+  for (int i = 0; i < dataSize; ++i) {
+    REQUIRE(out[i] == i * 2.0f);
+  }
+}
+
+TEST_CASE("usm_in_order_queue", "in_order_queue_solution") {
+  constexpr size_t dataSize = 1024;
+
+  float inA[dataSize], inB[dataSize], inC[dataSize], out[dataSize];
+  for (int i = 0; i < dataSize; ++i) {
+    inA[i] = static_cast<float>(i);
+    inB[i] = static_cast<float>(i);
+    inC[i] = static_cast<float>(i);
+    out[i] = 0.0f;
+  }
+
+  try {
+    auto asyncHandler = [&](sycl::exception_list exceptionList) {
+      for (auto& e : exceptionList) {
+        std::rethrow_exception(e);
+      }
+    };
+
+    auto inOrderQueue = sycl::queue{
+        usm_selector{}, asyncHandler, {sycl::property::queue::in_order{}}};
+#ifdef SYCL_ACADEMY_USING_COMPUTE_CPP
+    auto devicePtrInA = sycl::experimental::usm_wrapper<float>{
+        sycl::malloc_device<float>(dataSize, inOrderQueue)};
+    auto devicePtrInB = sycl::experimental::usm_wrapper<float>{
+        sycl::malloc_device<float>(dataSize, inOrderQueue)};
+    auto devicePtrInC = sycl::experimental::usm_wrapper<float>{
+        sycl::malloc_device<float>(dataSize, inOrderQueue)};
+    auto devicePtrOut = sycl::experimental::usm_wrapper<float>{
+        sycl::malloc_device<float>(dataSize, inOrderQueue)};
+#else
+    auto devicePtrInA = sycl::malloc_device<float>(dataSize, inOrderQueue);
+    auto devicePtrInB = sycl::malloc_device<float>(dataSize, inOrderQueue);
+    auto devicePtrInC = sycl::malloc_device<float>(dataSize, inOrderQueue);
+    auto devicePtrOut = sycl::malloc_device<float>(dataSize, inOrderQueue);
+#endif
+
+    inOrderQueue.memcpy(devicePtrInA, inA, sizeof(float) * dataSize);
+    inOrderQueue.memcpy(devicePtrInB, inB, sizeof(float) * dataSize);
+    inOrderQueue.memcpy(devicePtrInC, inC, sizeof(float) * dataSize);
+    inOrderQueue.memcpy(devicePtrOut, out, sizeof(float) * dataSize);
+
+    inOrderQueue
+        .parallel_for<kernel_a_2>(sycl::range{dataSize},
+                                  [=](sycl::id<1> idx) {
+                                    auto globalId = idx[0];
+                                    devicePtrInA[globalId] =
+                                        devicePtrInA[globalId] * 2.0f;
+                                  })
+        .wait();
```

Comment:
These waits shouldn't be necessary with an in-order queue

Reply:
Yeah that's true, not sure why we were calling `wait` here.